### DISCUSSION TITLE
Provide a way to disable table of contents in .mdx files

### DISCUSSION THREAD

**Discussion by nlopin** [storybookjs/storybook#23672] [Ideas]
### Is your feature request related to a problem? Please describe.

If Storybook is used to create a documentation site, it contains a mix of files with stories and unattached mdx pages.
Automatically generated table of contents is really helpful, but if it is enabled for all the pages, it is not possible to disable in MDX files. 

This can lead to situations, when the page layout is broken:
![CleanShot 2023-08-01 at 19 30 07@2x](https://github.com/storybookjs/storybook/assets/1469636/7e8ed66b-3d81-43e6-aecb-3bd006708841)

### Describe the solution you'd like

Provide `parameters` or a subset of it to the `<Meta>` component in MDX the same way I can provide parameters in CSF files:

```
<Meta title='Welcome' toc={{disable: true}} />
```

### Are you able to assist to bring the feature to reality?

yes, I can

**Comment by noranda** [comment]
Bumping this. I would absolutely love to use this feature in my story docs and certain unattached docs with no stories, but I cannot simply because it's either all or nothing. I have some guides that don't need it and will break the layout even if there are no headers. This is a fantastic feature, but it needs more control!

**Reply by nlopin** [reply]
I ended up hiding the TOC on individual pages via CSS 😢

**Reply by noranda** [reply]
I was playing around with this again yesterday and was able to do the same, only by absolutely positioning my top level div and giving it an inset. The empty toc with a set width is still there but now hidden. A hacky solution but I'll take it!

**Comment by Christo44** [comment]
how do you do this with css ? I don't see any unique selectors to do it per story.

I also tried to write a script in preview-head.html which waits for the DOM, then looks at the window path to apply a class to the preview body or iframe..

```
document.addEventListener('DOMContentLoaded', function () {
```
which seems overkill.

**Reply by noranda** [reply]
There wasn't a great way to do this. I have a welcome page with a background color, so I just absolutely positioned the entire page with Tailwind. In my MDX file I render this component:
```jsx
import {Unstyled} from '@storybook/blocks';
// ...
<Unstyled>
  <div className="absolute inset-0 flex overflow-hidden bg-blue">
    {/* ... */}
  </div>
</Unstyled>
```
The TOC is still there, but since I have no header elements, it's just an empty div and is hidden behind my component div.

Hope this helps! Would still love to see better support for this.

**Comment by Christo44** [comment]
I'm not familiar with the unstyled block. Ill give it a look.

My problem may be that I'm using vite/vue builder. And I'm not using blocks,  Im using mdx to write my stories.

**Reply by sirchnik-x** [reply]
This works for me with vite/vue.
```mdx
<style>
  {`
  .toc-wrapper {
    display: none;
  }
  `}
</style>
```

**Reply by noranda** [reply]
If it helps, you can use blocks in your MDX files, which is what I do. `Unstyled` just removes all of Storybook's base story styles, which can conflict with your styles.

**Reply by nedkelly** [reply]
If you find this doesn't quite do the job:

> This works for me with vite/vue.
> 
> ```mdx
> <style>
>   {`
>   .toc-wrapper {
>     display: none;
>   }
>   `}
> </style>
> ```

This worked for me:

```mdx
<style>
  {`
    .sbdocs-wrapper > div:has(div > .toc-wrapper) {
      display: none;
    }
  `}
</style>
```

**Reply by swapnilmmane** [reply]
For the MDX file this worked for me as well, thanks @nedkelly!

```
<style>
  {`
    .sbdocs-wrapper > div:has(div > .toc-wrapper) {
      display: none;
    }
  `}
</style>
```

**Comment by bromanazzi** [comment]
I made a selector so that the `toc` is only enabled for autodocs. Suits our usecase:

in `.storybook/preview-head.html:`

```html
<style>
  .sbdocs-wrapper:not(:has(.autodocs)) > div:has(div > .toc-wrapper) {
      display: none;
  }
</style>
```

For it to work we need to add a div with a class of autodocs to the template:

```js
page: () => (
    <>
      <div className="autodocs" style={{ display: "none" }} />
      <Title />
      <Subtitle />
      <Description />
      <Primary />
      <Stories />
      <Controls />
    </>
  ),
```

I suppose a similar approach could be used for discriminating mdx files. Not ideal but works for us.

**Comment by aimerie** [comment]
I would just love to see a better ability to customize the toc options on a mdx file. The specific scenario I want to support is customizing the header selection to include an additional level on the .mdx pages that I don't need elsewhere on the pages generated with autodocs.

**Comment by nickvisut** [comment]
Looks like somebody may have solved this problem by getting `<Meta/>` to recognize the `toc` prop through some `main.ts` configuration, described in the comments of another issue: https://github.com/storybookjs/storybook/issues/26521#issuecomment-2536880666. Perhaps someone could confirm that it behaves similarly to [what the OP originally asked for](#discussion-5464682)?

---

However, I have a slightly different use case where I'd like to hide any _empty_ table of contents because they interrupt the layout as pictured:
<img width="1177" height="541" alt="empty-toc-layout-issue" src="https://github.com/user-attachments/assets/c32bef88-e9cb-4247-9493-e84aacbb8706" />

Leaving this CSS snippet workaround here for anyone with the same use case ([put it in a CSS file and import into `preview.tsx`](https://storybook.js.org/docs/configure/styling-and-css#import-bundled-css-recommended)):

```css
/* Hide TOC if there is no content */
/* NOTE: alternatively, use `.sbdocs.sbdocs-toc--custom:not(:has(.toc-list))` which may be slightly less performant, but more succinct and arguably less dependent on the exact HTML structure */
.sbdocs.sbdocs-toc--custom:not(:has(> nav > .toc-wrapper > .toc-list)) {
  display: none;
}
```

You could probably also put it in [previewHead](https://storybook.js.org/docs/10/api/main-config/main-config-preview-head) or [preview-head.html](https://storybook.js.org/docs/9/configure/story-rendering#adding-to-head), as mentioned by others, although the docs note that the latter doesn't work with HMR.
